libgit/autogit_node_wrappers.go

```diff
@@ -0,0 +1,174 @@
+	tlfType tlf.Type
+}
+
+var _ libkbfs.Node = (*tlfTypeNode)(nil)
+
+// ShouldCreateMissedLookup implements the Node interface for
+// tlfTypeNode.
+func (ttn tlfTypeNode) ShouldCreateMissedLookup(
+	ctx context.Context, name string) (
+	bool, context.Context, libkbfs.EntryType, string) {
+	_, err := libkbfs.ParseTlfHandle(
+		ctx, ttn.am.config.KBPKI(), ttn.am.config.MDOps(), name, ttn.tlfType)
+
+	ctx = context.WithValue(ctx, ctxReadWriteKey, 1)
+	switch e := errors.Cause(err).(type) {
+	case nil:
+		return true, ctx, libkbfs.Dir, ""
+	case libkbfs.TlfNameNotCanonical:
+		return true, ctx, libkbfs.Sym, e.NameToTry
+	default:
+		ttn.am.log.CDebugf(ctx,
+			"Error parsing handle for name %s: %+v", name, err)
+		return ttn.Node.ShouldCreateMissedLookup(ctx, name)
+	}
+}
+
+// autogitRootNode represents the .kbfs_autogit folder, and can only
+// contain subdirectories corresponding to TLF types.
+type autogitRootNode struct {
+	libkbfs.Node
+	am *AutogitManager
+}
+
+var _ libkbfs.Node = (*autogitRootNode)(nil)
+
+// ShouldCreateMissedLookup implements the Node interface for
+// autogitRootNode.
+func (arn autogitRootNode) ShouldCreateMissedLookup(
+	ctx context.Context, name string) (
+	bool, context.Context, libkbfs.EntryType, string) {
+	switch name {
+	case public, private, team:
+		ctx = context.WithValue(ctx, ctxReadWriteKey, 1)
+		return true, ctx, libkbfs.Dir, ""
+	default:
+		return arn.Node.ShouldCreateMissedLookup(ctx, name)
+	}
+}
+
+// WrapChild implements the Node interface for autogitRootNode.
+func (arn autogitRootNode) WrapChild(child libkbfs.Node) libkbfs.Node {
+	child = arn.Node.WrapChild(child)
+	var tlfType tlf.Type
+	switch child.GetBasename() {
+	case public:
+		tlfType = tlf.Public
+	case private:
+		tlfType = tlf.Private
+	case team:
+		tlfType = tlf.SingleTeam
+	default:
+		return child
+	}
+	return &tlfTypeNode{child, arn.am, tlfType}
+}
+
+// readonlyNode is a read-only node by default, unless `ctxReadWriteKey`
+// has a value set in the context.
+type readonlyNode struct {
+	libkbfs.Node
+}
+
+var _ libkbfs.Node = (*readonlyNode)(nil)
+
+// Readonly implements the Node interface for readonlyNode.
+func (rn readonlyNode) Readonly(ctx context.Context) bool {
+	return ctx.Value(ctxReadWriteKey) == nil
+}
+
+// WrapChild implements the Node interface for readonlyNode.
+func (rn readonlyNode) WrapChild(child libkbfs.Node) libkbfs.Node {
+	return &readonlyNode{rn.Node.WrapChild(child)}
+}
+
+// rootNode is a Node wrapper around a TLF root node, that causes the
+// autogit root to be created when it is accessed.
+type rootNode struct {
+	libkbfs.Node
+	am *AutogitManager
+}
+
+var _ libkbfs.Node = (*rootNode)(nil)
+
+// ShouldCreateMissedLookup implements the Node interface for
+// rootNode.
+func (rn rootNode) ShouldCreateMissedLookup(ctx context.Context, name string) (
+	bool, context.Context, libkbfs.EntryType, string) {
+	if name == autogitRoot {
+		ctx = context.WithValue(ctx, ctxReadWriteKey, 1)
```

Comment:
It just needs to have the key/value in context right? Perhaps make the value type `struct{}` to avoid confusion (where one may think the value matters). Same for above where this ctx key is used.